File: 4-median-of-two-sorted-arrays/median-of-two-sorted-arrays.cpp
#include <vector>
#include <algorithm>
#include <climits>
using namespace std;

class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        // Make sure nums1 is the smaller array to optimize binary search
        if (nums1.size() > nums2.size())
            return findMedianSortedArrays(nums2, nums1);
        
        int m = nums1.size();
        int n = nums2.size();
        int total = m + n;
        int half = (total + 1) / 2;
        
        int left = 0, right = m;
        
        while (left <= right) {
            int i = left + (right - left) / 2;  // partition index for nums1
            int j = half - i;                    // partition index for nums2
            
            int nums1_left = (i == 0) ? INT_MIN : nums1[i - 1];
            int nums1_right = (i == m) ? INT_MAX : nums1[i];
            
            int nums2_left = (j == 0) ? INT_MIN : nums2[j - 1];
            int nums2_right = (j == n) ? INT_MAX : nums2[j];
            
            if (nums1_left <= nums2_right && nums2_left <= nums1_right) {
                // Correct partition found
                if (total % 2 == 1) {
                    return double(max(nums1_left, nums2_left));
                } else {
                    return (max(nums1_left, nums2_left) + min(nums1_right, nums2_right)) / 2.0;
                }
            } else if (nums1_left > nums2_right) {
                right = i - 1;
            } else {
                left = i + 1;
            }
        }
        
        // If input arrays are valid, this line will never be reached
        return 0.0;
    }
};

File: 4-median-of-two-sorted-arrays/median-of-two-sorted-arrays.py
class Solution(object):
    def findMedianSortedArrays(self, nums1, nums2):
        # Ensure nums1 is the smaller array for binary search efficiency
        if len(nums1) > len(nums2):
            nums1, nums2 = nums2, nums1

        m, n = len(nums1), len(nums2)
        total_len = m + n
        half_len = (total_len + 1) // 2  # +1 ensures correct handling for odd lengths

        left, right = 0, m
        while left <= right:
            i = (left + right) // 2  # Partition in nums1
            j = half_len - i         # Partition in nums2

            # Get the elements just before and after the partition
            max_left1 = float('-inf') if i == 0 else nums1[i-1]
            min_right1 = float('inf') if i == m else nums1[i]

            max_left2 = float('-inf') if j == 0 else nums2[j-1]
            min_right2 = float('inf') if j == n else nums2[j]

            # Check if we have a valid partition
            if max_left1 <= min_right2 and max_left2 <= min_right1:
                # Found the correct partition
                if total_len % 2 == 0:
                    return (max(max_left1, max_left2) + min(min_right1, min_right2)) / 2.0
                else:
                    return float(max(max_left1, max_left2))
            elif max_left1 > min_right2:
                right = i - 1  # Too far right in nums1, move left
            else:
                left = i + 1   # Too far left in nums1, move right
